src/io/linux.zig

```diff
@@ -1098,8 +1157,66 @@ pub const IO = struct {
 
-        const stat = try os.fstat(fd);
-        if (stat.size < size) @panic("data file inode size was truncated or corrupted");
+        switch (kind) {
+            .file => {
+                if ((try os.fstat(fd)).size < size) {
+                    @panic("data file inode size was truncated or corrupted");
+                }
+            },
+            .block_device => {
+                const BLKGETSIZE64 = os.linux.IOCTL.IOR(0x12, 114, usize);
+                var block_device_size: usize = 0;
+
+                switch (std.os.errno(os.linux.ioctl(
+                    fd,
+                    BLKGETSIZE64,
+                    @intFromPtr(&block_device_size),
+                ))) {
+                    .SUCCESS => {},
+
+                    // These are the only errors that are supposed to be possible from ioctl(2).
+                    .BADF => return error.InvalidFileDescriptor,
+                    .NOTTY => return error.BadRequest,
+                    .FAULT => return error.InvalidAddress,
+                    else => |err| return os.unexpectedErrno(err),
+                }
+
+                if (block_device_size < size) {
+                    std.debug.panic(
+                        "The block device used is too small ({} needed/{} available).",
+                        .{
+                            std.fmt.fmtIntSizeBin(size),
+                            std.fmt.fmtIntSizeBin(block_device_size),
+                        },
+                    );
+                }
+
+                if (method == .create or method == .create_or_open) {
+                    // Check that the first superblock_zone_size bytes are 0.
+                    // - It'll ensure that the block device is not directly TigerBeetle.
+                    // - It'll be very likely to catch any cases where there's an existing
+                    //   other filesystem.
+                    // - In the case of there being a partition table (eg, two partitions,
+                    //   one starting at 0MB, one at 1024MB) and the operator tries to format
+                    //   the raw disk (/dev/sda) while a partition later is
+                    //   TigerBeetle (/dev/sda2) it'll be blocked by the MBR/GPT existing.
+                    const superblock_zone_size = @import("../vsr/superblock.zig").superblock_zone_size;
+                    var read_buf: [superblock_zone_size]u8 align(constants.sector_size) = undefined;
+
+                    // We can do this without worrying about retrying partial reads because on linux,
+                    // read(2) on block devices can not be interrupted by signals. See signal(7).
+                    assert(superblock_zone_size == try os.read(fd, &read_buf));
+                    if (!std.mem.allEqual(u8, &read_buf, 0)) {
+                        @panic(
+                            "Superblock on block device not empty. If this is the correct block device to use, please format it first.",
```

Comment:
Could we make this print out the `dd` command to zero the device? Alternatively, if we don't want to risk that, perhaps change the wording to something like

```suggestion
                            "Superblock on block device not empty. If this is the correct block device to use, please zero the first {}MB using a tool like dd.",
```

(since format might be confusing since the user is running `tigerbeetle format`, and formatting it with eg `mkfs` won't work either)